salt/modules/nxos.py

```diff
@@ -57,11 +225,636 @@ def cmd(command, *args, **kwargs):
+    else:
+        roles = []
+    return roles
+
+
+def get_user(username, **kwargs):
+    """
+    Get username line from switch.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd get_user username=admin
+    """
+    command = 'show run | include "^username {0} password 5 "'.format(username)
+    info = ""
+    info = show(command, **kwargs)
+    if isinstance(info, list):
+        info = info[0]
+    return info
+
+
+def grains(**kwargs):
+    """
+    Get grains for minion.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd grains
+    """
+    if not DEVICE_DETAILS["grains_cache"]:
+        ret = salt.utils.nxos.system_info(show_ver(**kwargs))
+        log.debug(ret)
+        DEVICE_DETAILS["grains_cache"].update(ret["nxos"])
+    return DEVICE_DETAILS["grains_cache"]
+
+
+def grains_refresh(**kwargs):
+    """
+    Refresh the grains for the NX-OS device.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd grains_refresh
+    """
+    DEVICE_DETAILS["grains_cache"] = {}
+    return grains(**kwargs)
+
+
+def sendline(command, method="cli_show_ascii", **kwargs):
+    """
+    Send arbitray commands to the NX-OS device.
+
+    command
+        The command to be sent.
+
+    method:
+        ``cli_show_ascii``: Return raw test or unstructured output.
+        ``cli_show``: Return structured output.
+        ``cli_conf``: Send configuration commands to the device.
+        Defaults to ``cli_show_ascii``.
+        NOTE: method is ignored for SSH proxy minion.  All data is returned
+        unstructured.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd sendline 'show run | include "^username admin password"'
+    """
+    smethods = ["cli_show_ascii", "cli_show", "cli_conf"]
+    if method not in smethods:
+        msg = """
+        INPUT ERROR: Second argument 'method' must be one of {0}
+        Value passed: {1}
+        Hint: White space separated commands should be wrapped by double quotes
+        """.format(
+            smethods, method
+        )
+        return msg
+
+    try:
+        if salt.utils.platform.is_proxy():
+            return __proxy__["nxos.sendline"](command, method, **kwargs)
+        else:
+            return _nxapi_request(command, method, **kwargs)
+    except socket_error as e:
+        return e.strerror + "\n" + CONNECTION_ERROR_MSG
+    except NxosError as e:
+        return e.strerror + "\n" + CONNECTION_ERROR_MSG
+
+
+def show(commands, raw_text=True, **kwargs):
+    """
+    Execute one or more show (non-configuration) commands.
+
+    commands
+        The commands to be executed.
+
+    raw_text: ``True``
+        Whether to return raw text or structured data.
+        NOTE: raw_text option is ignored for SSH proxy minion.  Data is
+        returned unstructured.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt-call --local nxos.show 'show version'
+        salt '*' nxos.show 'show bgp sessions ; show processes' raw_text=False
```

Comment:
The repeating "show" seems weird. 

Also, the default of raw_text is the same default as for sendline, so the flag seems less important?

Reply:
This is legacy behavior.  See:

https://github.com/saltstack/salt/blob/master/salt/modules/nxos_api.py#L300

Reply:
Should we modify salt/modules/nxos_api.py to call the functions in modules/nxos?

Reply:
I don't really feel comfortable modifying the functions in `nxos_api` and also believe that work is beyond the scope of what we are doing here.  I would prefer to mark the entire `nxos_api` module for deprecation if that is possible.

Reply:
Sure, we can do that :)

Reply:
`napalm` also uses `nxos_api`, should we update it to use the equivalent `nxos` functions?

Reply:
@dhiltonp I was not aware of the fact that `napalm` was using `nxos_api`.  In that case maybe we don't want to deprecate `nxos_api`.  I do believe that `napalm` should be updated at some point to use the `nxos` functions but that should involve the `napalm` folks and is beyond the scope of this PR.